sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java

```diff
@@ -65,6 +66,16 @@ private Reshuffle() {}
     return new ViaRandomKey<>();
   }
 
+  @Experimental
```

Comment:
Annotation should probably be below the comment.

Also suggest rephrasing a bit, to explain when one should use one or the other, and what are the consequences of choosing wrong. I'm not sure that "sequentially generated" is clear enough to a casual user. Maybe something like this:

```
Materializes the input and prepares it to be consumed in a highly parallel fashion.

This version is tailored to the case when input was produced in an extremely sequential way - typically by a ParDo that emits millions of outputs _per input element_, e.g., executing a large database query or a large simulation and emitting all of their results.

Internally, this version first materializes the input at a moderate cost before reshuffling it internally using viaRandomKey(), making the reshuffling itself significantly cheaper in these extreme cases on some runners. Use this over viaRandomKey() only if your benchmarks show an improvement.
```

And mention this at the class-level documentation of Reshuffle for visibility.

Reply:
I think the key is that the input was generated with limited parallelism and suggest that we update `ViaRandomKey` to perform both expansions where the side input fusion break is added if the `withHintLimitedInputParallelism()` is specified.

Reply:
I think this (`withHint...`) is a great idea, thanks Luke. It also wouldn't hurt to rerun the benchmarks (and to check them in - sadly I didn't do it back then, and probably don't have them anymore, but it only took like 15 min to write).